File: Leetcode2429.py
class Solution:
    def minimizeXor(self, num1: int, num2: int) -> int:
        goal = 0
        set_bits = 0
        while(num2 != 0):
            if num2%2:
                set_bits += 1
            num2 >>= 1
        temp_num1 = num1 >> 1
        max_val = 1
        while(temp_num1>0):
            max_val *= 2
            temp_num1 >>= 1
        while(set_bits>0 and max_val>0):
            if max_val&num1:
                goal += max_val
                set_bits -= 1
            max_val >>= 1
        addend = 1
        while(set_bits>0):
            if num1%2 == 0:
                goal += addend
                set_bits -= 1
            addend *= 2
            num1 >>= 1
        return goal
        

File: Leetcode2429.cpp
class Solution {
public:
    int minimizeXor(int num1, int num2) {
        int minimized = 0;
        int setBits = 0;
        while(num2 > 0){
            setBits += num2 % 2;
            num2 /= 2;
        }
        int base = 0;
        int num1Clone = num1;
        while(num1Clone > 1){
            base++;
            num1Clone /= 2;
        }
        while(num1 > 0 && setBits > 0){
            int currentPower = pow(2, base);
            if(currentPower & num1){
                minimized += currentPower;
                setBits--;
            }
            num1 %= currentPower;
            base--;
        }
        base = 0;
        while(setBits > 0){
            int currentPower = pow(2, base);
            if(!(minimized & currentPower)){
                minimized += currentPower;
                setBits--;
            }
            base++;
        }
        return minimized;
    }
};
